### PR TITLE
fix: should not tick nested timer in fakeAsync test

### LINKED ISSUE #33799: FakeAsync's tick flushs nested timeouts
# 🐞 bug report

### Affected Package
The issue is caused by package @angular/core/testing

### Is this a regression?

dunno

### Description

`tick` call flushs even nested timeouts that have not been installed at the call moment

## 🔬 Minimal Reproduction

```ts
function waitTwice(callback: () => any): void {
  window.setTimeout(() =>
    window.setTimeout(() => callback())
  );
}

describe('fakeAsync', () => {
  it('nested timers', fakeAsync(() => {
    const callback = jasmine.createSpy('callback');

    waitTwice(callback);

    expect(callback).not.toHaveBeenCalled();

    tick();

    expect(callback).not.toHaveBeenCalled();
  }));
});
```

repo with this test

- `git clone git@github.com:artaommahe/fakeasync-nested-timeout.git`
- `yarn install`
- `yarn test`

[test file](https://github.com/artaommahe/fakeasync-nested-timeout/blob/master/src/app/fakeasync.spec.ts)

## 🔥 Exception or Error
```
Chrome 78.0.3904 (Windows 10.0.0) fakeAsync nested timers FAILED
 Error: Expected spy callback not to have been called.
```

## 🌍  Your Environment

**Angular Version:**
<pre><code>

Angular CLI: 8.3.18
Node: 10.16.1
OS: win32 x64
Angular: 8.2.13
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.803.18
@angular-devkit/build-angular     0.803.18
@angular-devkit/build-optimizer   0.803.18
@angular-devkit/build-webpack     0.803.18
@angular-devkit/core              8.3.18
@angular-devkit/schematics        8.3.18
@angular/cli                      8.3.18
@ngtools/webpack                  8.3.18
@schematics/angular               8.3.18
@schematics/update                0.803.18
rxjs                              6.4.0
typescript                        3.5.3
webpack                           4.39.2

</code></pre>

**Anything else relevant?**

Same test works well with jest's timer functions

```ts
jest.useFakeTimers();
const callback = jest.fn();
waitTwice(callback);

jest.runOnlyPendingTimers();

expect(callback).not.toHaveBeenCalled();

jest.runOnlyPendingTimers();

expect(callback).toHaveBeenCalled();
```

It is a real world case (function with nested zero-timeouts) - we are forced to use this due to parent-child init hooks order for custom elements and for manual elements creation (creating components tree from string markup) when parent's ngAfterViewInit is called sooner than child's one.

### PR DESCRIPTION
**Pull request by JiaLiPassion**
Close #33799

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #33799
```
function waitTwice(callback: () => any): void {
  setTimeout(() =>
    setTimeout(() => callback())
  );
}

describe('fakeAsync', () => {
  it('nested timers', fakeAsync(() => {
    const callback = jasmine.createSpy('callback');
    waitTwice(callback);
    expect(callback).not.toHaveBeenCalled();
    tick();
    expect(callback).toHaveBeenCalled();
    tick();
    expect(callback).toHaveBeenCalled();
  }));
});
```

By default, 
The `callback` will be called after the `first tick`.

```
describe('fakeAsync', () => {
  it('nested timers', fakeAsync(() => {
    const callback = jasmine.createSpy('callback');
    waitTwice(callback);
    expect(callback).not.toHaveBeenCalled();
    tick(0, true);
    expect(callback).not.toHaveBeenCalled();
    tick();
    expect(callback).toHaveBeenCalled();
  }));
});
```
````
## What is the new behavior?
with `tick(0, true)`, the second parameter is `ignoreNestedTimeout`, after the `first tick`, the `nested timeout` will not be invoked.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:281123086:BASE:281123154:1574107580499:df5f311f)

File: packages/zone.js/lib/zone-spec/fake-async-test.ts [hunk 1]
```diff
@@ -83,7 +85,8 @@
 
     scheduleFunction(
         cb: Function, delay: number, args: any[] = [], isPeriodic: boolean = false,
-        isRequestAnimationFrame: boolean = false, id: number = -1): number {
+        isRequestAnimationFrame: boolean = false, id: number = -1,
```
Comment (mhevery) [inline packages/zone.js/lib/zone-spec/fake-async-test.ts]: these function args are getting too long and `id` is inside of the `isRequestAnimationFrame` and `isRequeuePeriodic`. What do you think about refactoring it to have optional named arguments instead?

**Comment by mhevery** [comment]
So this change is breaky in `g3` where 30+ targets fail just in blueprint.

**Review by mhevery — COMMENTED** [review]
could we add some API docs to the change?

**Review by petebacondarwin — COMMENTED** [review]
The last "`fix: format`" commit should be a fixup commit for the previous commit

**Comment by JiaLiPassion** [comment]
@mhevery , @petebacondarwin , thank you for the review, I have updated the PR, use new API instead of breaking changes. Please review again, thanks!

**Review by petebacondarwin — APPROVED** [review]
LGTM

File: packages/zone.js/lib/zone-spec/fake-async-test.ts (hunk 1 above)
Comment (JiaLiPassion) [reply packages/zone.js/lib/zone-spec/fake-async-test.ts]: @mhevery , yeah, you are right, in the updated PR, I refactored the parameters to optional named arguments object.

File: packages/core/testing/src/fake_async.ts [hunk 2]
```diff
@@ -62,13 +62,17 @@ export function fakeAsync(fn: Function): (...args: any[]) => any {
...
  * {@example core/testing/ts/fake_async.ts region='basic'}
  *
+ * @param millis, the number of millisecond to advance the virtual timer
+ * @param ignoreNestedTimeout, whether to invoke the nested timeout, by default is false,
+ *                             means the nested timeout will be triggered
+ *
  * @publicApi
  */
-export function tick(millis: number = 0): void {
+export function tick(millis: number = 0, ignoreNestedTimeout = false): void {
```
Comment (mhevery) [inline packages/core/testing/src/fake_async.ts]: I fear that over time we will have more of these arguments. Could we turn the options into object literal? That way adding new options in the future will not be an issue.?

File: packages/core/testing/src/fake_async.ts [hunk 3]
```diff
@@ -62,13 +62,17 @@ export function fakeAsync(fn: Function): (...args: any[]) => any {
  *
  * {@example core/testing/ts/fake_async.ts region='basic'}
  *
+ * @param millis, the number of millisecond to advance the virtual timer
+ * @param ignoreNestedTimeout, whether to invoke the nested timeout, by default is false,
```
Comment (mhevery) [inline packages/core/testing/src/fake_async.ts]: Could you add more text here. Do you mean if the nested timeout will be flushed in the same tick or separately?

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:292587640:BASE:292587675:1580500248141:56894f3f)

File: packages/core/testing/src/fake_async.ts (hunk 2 above)
Comment (mhevery) [reply packages/core/testing/src/fake_async.ts]: Also I think the name should be `ignoreNestedTimeout` => `processNewMacroTasksSynchronously` (Open to other names)  What do you think. If I understand correctly, this is not specific to `setTimeout` but rather to the fact that if a macro task is scheduled in the current flush cycle, wether or not it should be flushed synchronously.

**Review by mhevery — COMMENTED** [review]
Minor suggestions. Running presubmit now.

File: packages/core/testing/src/fake_async.ts (hunk 2 above)
Comment (JiaLiPassion) [reply packages/core/testing/src/fake_async.ts]: yes, you are right, it is not only `setTimeout`, the `processNewMacroTasksSynchronously` is a better name, thanks.

File: packages/core/testing/src/fake_async.ts (hunk 3 above)
Comment (JiaLiPassion) [reply packages/core/testing/src/fake_async.ts]: Yes, I have added more text and sample codes. Thanks.

**Review by mhevery — APPROVED** [review]

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:292587640:BASE:294463027:1581444275652:c9c782ac)

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:292587640:BASE:296256842:1582226671611:88bd667)

**Pull request closed without merging**